src/java.base/share/classes/java/lang/StringConcatHelper.java

```diff
@@ -119,91 +119,64 @@ static long mix(long lengthCoder, long value) {
      */
     static long mix(long lengthCoder, String value) {
         lengthCoder += value.length();
-        if (value.coder() == String.UTF16) {
+        if (!value.isLatin1()) {
             lengthCoder |= UTF16;
         }
         return checkOverflow(lengthCoder);
     }
 
     /**
-     * Prepends the stringly representation of boolean value into buffer,
+     * Prepends constant and the stringly representation of value into buffer,
      * given the coder and final index. Index is measured in chars, not in bytes!
      *
      * @param indexCoder final char index in the buffer, along with coder packed
      *                   into higher bits.
      * @param buf        buffer to append to
      * @param value      boolean value to encode
+     * @param prefix     a constant to prepend before value
      * @return           updated index (coder value retained)
      */
-    static long prepend(long indexCoder, byte[] buf, boolean value) {
+    static long prepend(long indexCoder, byte[] buf, boolean value, String prefix) {
         int index = (int)indexCoder;
         if (indexCoder < UTF16) {
             if (value) {
-                buf[--index] = 'e';
-                buf[--index] = 'u';
-                buf[--index] = 'r';
-                buf[--index] = 't';
+                index -= 4;
+                buf[index] = 't';
+                buf[index + 1] = 'r';
+                buf[index + 2] = 'u';
+                buf[index + 3] = 'e';
             } else {
-                buf[--index] = 'e';
-                buf[--index] = 's';
-                buf[--index] = 'l';
-                buf[--index] = 'a';
-                buf[--index] = 'f';
+                index -= 5;
+                buf[index] = 'f';
+                buf[index + 1] = 'a';
+                buf[index + 2] = 'l';
+                buf[index + 3] = 's';
+                buf[index + 4] = 'e';
             }
+            index -= prefix.length();
+            prefix.getBytes(buf, index, String.LATIN1);
```

Comment:
Since we are now passing in a lot of empty prefix, I wonder how this call is elided; is there some specific JIT intrinsic? The java code has no shortcut for `length == 0`.

Reply:
Remember that the prefixes are all constants and bound into the MH at each call site, so one view is that this PR is mostly about tickling the code into something that just-so happens to be easier for the JIT to swallow